### PR TITLE
Issue 885 - `LayoutProvider.deleteMetadata` method should be moved into `Maven2FileSystemProvider

### LINKED ISSUE #885: `LayoutProvider.deleteMetadata` method should be moved into `Maven2FileSystemProvider`
What need to be done:
- move `deleteMetadata` method from `LayoutProvider` into `RepositoryLayoutFileSystemProvider` with `protected` modifier
- `RepositoryLayoutFileSystemProvider.deleteMetadata` should be called within `RepositoryLayoutFileSystemProvider.delete`, so all corresponding metadata should be removed within artifact delete,
- move `Maven2LayoutProvider.deleteMetadata` logic into `Maven2FileSystemProvider.deleteMetadata`
- use `Files.delete` method instead of `LayoutProvider.deleteMetadata` in `MavenMetadataGroupRepositoryComponent.cleanupGroupW
 henArtifactPathNoLongerExistsInSubTree`
- check that corresponding logic in `GroupRepositoryArtifactExistenceChecker` covered with Unit Tests
- provide additional tests if needed

### PR DESCRIPTION
**Pull request by adavid9**
I spoke with @carlspring he said that the guy probably left, so he asked me to rewrite this fix and rebase it with current master branch. This commit is fix issue #1037 which was related with `@Deprecated` method. 

Fixes #885

### CODE REVIEW

File: strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/LayoutFileSystemProvider.java [hunk 1]
```diff
@@ -372,4 +376,8 @@ public void undelete(RepositoryPath path)
         return getLayoutProvider().getRepositoryFileAttributes(repositoryRelativePath, attributeTypes);
     }
     
+    protected void deleteMetadata(RepositoryPath repositoryPath) throws IOException {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/providers/layout/LayoutFileSystemProvider.java]: Cosmetic: Braces and `throws` statements on new lines, please!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 2]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
         // Otherwise, this is either not an artifact directory, or not a valid Maven artifact
     }
 
-    
+    @Override
+    public void deleteMetadata(RepositoryPath artifactPath) {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Braces on new line, please!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 3]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+            {
+                //it's repository root directory, so we have nothing to clean here
+                return;
+            }
+
+            if (Files.exists(artifactPath))
+            {
+
+                RepositoryFileAttributes artifactFileAttributes = Files.readAttributes(artifactPath,
+                        RepositoryFileAttributes.class);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep method parameters aligned.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 4]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+
+                if (!artifactFileAttributes.isDirectory())
+                {
+                    artifactBasePath = artifactBasePath.getParent();
+                    artifactIdLevelPath = artifactIdLevelPath.getParent();
+
+                    // This is at the version level
+                    try (Stream<Path> pathStream = Files.list(artifactBasePath))
+                    {
+                        Path pomPath = pathStream.filter(
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep method parameters aligned.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 5]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+                                p -> p.getFileName().toString().endsWith(".pom")).findFirst().orElse(null);
+
+                        if (pomPath != null)
+                        {
+                            String version = ArtifactUtils.convertPathToArtifact(
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep method parameters aligned.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 6]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+                artifactBasePath = artifactBasePath.getParent();
+                artifactIdLevelPath = artifactIdLevelPath.getParent();
+            }
+
+            if (Files.exists(artifactIdLevelPath))
+            {
+                // This is at the artifact level
+                try (Stream<Path> pathStream = Files.list(artifactIdLevelPath))
+                {
+                    Path mavenMetadataPath = pathStream.filter(p -> p.getFileName()
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep method parameters aligned.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 7]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+        {
+            Metadata metadataVersionLevel = mavenMetadataManager.readMetadata(metadataBasePath);
+            if (metadataVersionLevel != null && metadataVersionLevel.getVersioning() != null &&
+                    metadataVersionLevel.getVersioning().getVersions().contains(version))
+            {
+                metadataVersionLevel.getVersioning().getVersions().remove(version);
+
+                MetadataHelper.setLastUpdated(metadataVersionLevel.getVersioning());
+
+                mavenMetadataManager.storeMetadata(metadataBasePath,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep chained method parameters aligned.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 8]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+                        null,
+                        metadataVersionLevel,
+                        MetadataType.SNAPSHOT_VERSION_LEVEL);
+            }
+        }
+    }
+
+    public void deleteMetadataAtArtifactLevel(RepositoryPath artifactPath,
+                                              String version) throws IOException, XmlPullParserException {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Could you please have the `throws` clause and braces on new lines?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java [hunk 9]
```diff
@@ -127,5 +135,130 @@ private void cleanupDirectory(RepositoryPath repositoryPathRelative,
...
+            }
+
+            if (version.equals(metadataVersionLevel.getVersioning().getRelease()))
+            {
+                MetadataHelper.setRelease(metadataVersionLevel);
+            }
+
+            MetadataHelper.setLastUpdated(metadataVersionLevel.getVersioning());
+
+            mavenMetadataManager.storeMetadata(artifactPath,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/providers/layout/Maven2FileSystemProvider.java]: Cosmetic: Please, keep chained method parameters aligned.

**Comment by carlspring** [comment]
@adavid9 :
> This commit is also fix for issue 1037 which was related with `@Deprecated` method.

Is this a partial or complete fix for #1037? If it's a complete fix, then could you please use the words `Fixes #1037`, so that Github automatically links and closes it, when this pull request is merged and closed? Thanks! :)

**Comment by carlspring** [comment]
@adavid9 ,

What is the status of this task? Is it ready for reviewing and merging?

**Comment by adavid9** [comment]
@carlspring Here is the fix https://github.com/strongbox/strongbox/pull/1039/commits/74f31b316a527fad4db642f6f95044a30daa0c39
As I wrote above this is also fix for #1037, you can merge it

**Comment by carlspring** [comment]
@adavid9 ,

There are still failing tests! Could you please fix them first, so I can merge it?

**Comment by carlspring** [comment]
@adavid9 ,

Would you mind rebasing, as this PR now has merge conflicts? Thanks!

@sbespalov ,

Could you please review this pull request, as it was based on a change requested by you? Thanks! :)

**Comment by adavid9** [comment]
@carlspring @sbespalov there are still failing tests, and honestly I am not sure why... @sbespalov if you will have some time could you please take a look at it? You was responsible for this code, so I believe that you will know much more than me, why it is failing.

**Comment by carlspring** [comment]
Thanks, @adavid9 ! :)

**Pull request merged**